pandas/core/dtypes/missing.py

```diff
@@ -223,15 +223,13 @@ def _isna_ndarraylike(obj, old: bool = False):
     array-like
         Array of boolean values denoting the NA status of each element.
     """
-    values = getattr(obj, "_values", obj)
+    if not isinstance(obj, np.ndarray):
+        values = obj.to_numpy()
+    else:
+        values = obj
     dtype = values.dtype
 
-    if is_extension_array_dtype(dtype):
-        if old:
-            result = values.isna() | (values == -np.inf) | (values == np.inf)
```

Comment:
I think the expression values == np.inf is actually a time bomb; the comparison ops for StringArray themselves call isna so you can end up hitting a RecursionError. Making sure we have a numpy array by this point and relying on the checks below seems like a safe way to fix it.

Reply:
Do you have an example where this gets triggered?

Reply:
It happens if you call isna on a StringArray backed Series within the use_inf_as_na context:
```python
import pandas as pd

pd.options.mode.use_inf_as_na = True

arr = pd.array(["a", "b", None])
ser = pd.Series(arr)
arr.isna() # Works
ser.isna() # Raises RecursionError
```
The StringArray case is fine because once we hit this we've already converted to numpy, but for Series we get here with a StringArray.

I've noticed that for IntegerArray at least we rely on the cached _mask attribute within the comparison ops and I think that would be very helpful to have here. That way it's at least permissible to do comparisons within these functions.

Reply:
You get the same with `pd.isna(arr)`.

But do you understand why we get a recursion error? 
`pd.isna(arr)` calls StringArray.isna, which calls `pd.isna` but on the underlying numpy array, so I don't see the infinite loop?

Reply:
Ah, or that is due to the equality comparison (`(values == -np.inf)`), since `StringArray.__eq__` also calls `isna(self)`. I suppose this could also be fixed by changing `isna(self)` to `isna(self._ndarray)` or `self.isna()` ?

https://github.com/pandas-dev/pandas/blob/d99b881873027604109e0b6b7935e144d0cc31b4/pandas/core/arrays/string_.py#L310

Reply:
I tried using isna(self._ndarray) here in the comparison op but was getting a weird test failure about a dtype mismatch, so eventually decided to go with isnaobj_old here since it's doing exactly what we want. Should a follow up be to look into this check within __eq__ to prevent future confusion / bugs?

Reply:
What error did you get?

>  so eventually decided to go with isnaobj_old here

it's maybe fine for now, but that does a potential conversion in `to_numpy`, which for some EAs would be nice to avoid

Reply:
One of the  tests added here was failing (the error message was for the dtype but actually the rest of the Series was incorrect):
```python
[ins] In [2]: pd.options.mode.use_inf_as_na = True                                                                                                                                                                                                                                                                                                                                                    

[ins] In [3]: arr = pd.array(["a", "b", "c"])                                                                                                                                                                

[ins] In [4]: ser = pd.Series(arr)                                                                                                                                                                           

[ins] In [5]: ser.isna()                                                                                                                                                                                     
Out[5]: 
0    <NA>
1    <NA>
2    <NA>
dtype: boolean
```
This is keeping the value == np.inf comparison but using self.isna() to calculate the mask inside the comparison op instead of isna(self). Would need to dig further to figure out what exactly is happening here.

Reply:
Does there exist, or would it be useful to have some kind of a cythonized helper function to check for identity with respect to a specific scalar value? Here for instance we want to efficiently check for equality with infinity but can't actually rely on ==. Could also be useful for generically checking which elements of an array-like are identically pd.NA. @jbrockmendel is this something like what you had in mind with the is_matching_na helper you'd mentioned a bit ago?

Reply:
Ah, that actually points to an interesting question: in `inf_as_na` mode, should also comparisons treat inf as na, and thus no longer actually compare inf?

```
In [8]: s = pd.Series([1.5, np.nan, np.inf]) 

In [9]: s == np.inf  
Out[9]: 
0    False
1    False
2     True
dtype: bool

In [10]:  pd.options.mode.use_inf_as_na = True    

In [11]: s == np.inf   
Out[11]: 
0    False
1    False
2    False
dtype: bool
```

I would *maybe* expect twice the same result ..

Reply:
My guess is that inf should become more or less indistinguishable from NaN under this option (at least I'm assuming that was the original intent here) and so it would behave as such in comparisons as well. You could argue philosophically that inf == inf already has dubious truth value 😄

Reply:
It should certainly become indistinguishable in "na" ops, like isna/dropna/fillna etc, but whether that also includes comparision ops, that's not directly clear to me. 

Now, this is rather a corner case, though, and we can always discuss it in a separate dedicated issue (if some feel strongly about it), and for now just assume the current behaviour. And you're fully correct that equality is already a strange concept for infinity ;)